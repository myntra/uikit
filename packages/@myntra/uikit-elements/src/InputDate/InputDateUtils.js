import PropTypes from 'prop-types'
import { parse as _parse, format as _format } from 'date-fns'

export const RE_ISO_DATE = /^(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d)$/

export const DateType = PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string])

export const DateRangeType = PropTypes.oneOfType([
  PropTypes.shape({
    from: PropTypes.string,
    to: PropTypes.string
  }),
  PropTypes.shape({
    from: PropTypes.instanceOf(Date),
    to: PropTypes.instanceOf(Date)
  })
])

const padStartWithZeros = (value, digits = 2) => `${value}`.padStart(digits, '0')

/**
 * Create date instance.
 *
 * @export
 * @param {number} year
 * @param {number} month
 * @param {number} date
 * @returns {Date}
 */
export function UTCDate(year, month, date) {
  return new Date(
    `${padStartWithZeros(year, 4)}-${padStartWithZeros(month + 1)}-${padStartWithZeros(date)}T00:00:00.000Z`
  )
}

/**
 * Drop time component from Date.
 *
 * @export
 * @param {Date} date
 * @returns {Date}
 */
export function onlyDate(date) {
  if (typeof date.toDate === 'function') {
    date = date.toDate() // dayjs interop.
  }

  return UTCDate(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate())
}

export function parse(str, format) {
  if (str instanceof Date) return str
  if (typeof str !== 'string' || str === '') return null
  if (RE_ISO_DATE.test(str)) return new Date(str)
  if (!format) throw new Error(`'format' is required.`)

  const date = _parse(str, format, new Date())

  if (date instanceof Date && !Number.isNaN(date.getTime())) {
    return UTCDate(date.getFullYear(), date.getMonth(), date.getDate())
  }

  throw new Error(`Date parse exception: '${str}' does not match '${format}'`)
}

export function format(value, pattern) {
  if (!value) return value
  if (!pattern) throw new Error(`'format' is required.`)
  if (value instanceof Date) return _format(value, pattern)
  if (typeof value === 'string') return format(parse(value, pattern), pattern)

  const result = {}
  if (value.from) result.from = format(value.from, pattern)
  if (value.to) result.to = format(value.to, pattern)
  return result
}

export function isDateEqual(a, b) {
  if (!(a instanceof Date)) return false
  if (!(b instanceof Date)) return false
  if (a.getTime() === b.getTime()) return true

  return (
    a.getUTCFullYear() === b.getUTCFullYear() &&
    a.getUTCMonth() === b.getUTCMonth() &&
    a.getUTCDate() === b.getUTCDate()
  )
}
