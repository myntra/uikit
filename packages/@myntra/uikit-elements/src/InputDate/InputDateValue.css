@import '../Input/Input.css';

.date-value {
  display: inline-flex;
  width: 100%;
}

.date-value > * {
  flex: 1;
}

.wrapper {
  position: relative;
}

.wrapper > .icon {
  cursor: pointer;
  position: absolute;
  right: var(--size-xx-small);
  bottom: var(--size-xx-small);
}

.date-value-active input {
  border-bottom: 1px solid var(--color-primary);
}

.active,
.active input {
  color: var(--color-primary);
}
