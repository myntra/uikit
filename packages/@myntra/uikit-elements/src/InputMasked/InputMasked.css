@import '@myntra/tokens/tokens.css';

.input {
  width: 100%;
  border: none;
  background-color: transparent;
  padding: var(--size-xx-small) var(--size-base) var(--size-xx-small) var(--size-none);
}

.mask {
  position: absolute;
  top: 0;
  left: 0;
  color: var(--color-gray-dark);
  pointer-events: none;
  z-index: -1;
}

.masked-input {
  border-bottom: 1px solid var(--color-gray-dark);
}

.masked-input:focus {
  outline: none;
  border-bottom: 1px solid var(--color-accent);
}

.masked-input:disabled {
  color: var(--typography-color-dark-secondary);
}

.input-container {
  position: relative;
}
