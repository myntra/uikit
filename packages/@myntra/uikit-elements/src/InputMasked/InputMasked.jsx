import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import defaultMasks from './masks'
import { classnames, each, map, findIndex, findLastIndex, memoize } from '@myntra/uikit-utils'

import styles from './InputMasked.css'
/**
 * Input component that provides a template for phone, credit card, etc. Mask patterns include
 * - d - A single digit from 0-9
 * - l - A lowercase letter
 * - L - An uppercase letter
 * - \* - A lowercase/uppercase letter
 * - A - An uppercase alphanumeric A-Z/0-9/_
 * - a - A lowercase alphanumeric a-z/0-9/_
 * - w - Any alphanumeric character A-Z/a-z/0-9/_
 * - Fixed Mask characters are placed inside double quotes, say, "Rs. "
 * @since 0.0.0
 * @status EXPERIMENTAL
 * @example
 * <InputMasked
 *   value={this.state.value}
 *   placeholder='PIN - XX XX'
 *   includeMaskChars={false}
 *   pattern='"PIN - "dd" "dd'
 *   onChange={(value) => this.setState({value})}
 * />
 */

export default class InputMasked extends PureComponent {
  static propTypes = {
    /** Value */
    value: PropTypes.string,
    /** Placeholder for the mask */
    placeholder: PropTypes.string,
    /** Include Mask characters in the value */
    includeMaskChars: PropTypes.bool,
    /** Mask pattern specified in the description */
    pattern: PropTypes.string,
    /** Additional mask definitions
     * @typedef {function} validateFunc
     * @param {string} char - character to validate
     * @return {bool} - is input valid
     *
     * @typedef {function} transformFunc
     * @param {string} char - character to transform
     * @return {string} - Transformed character
     *
     * @type {Object.<string, { validate: validateFunc, ?transform: transformFunc}>}
     */
    masks: PropTypes.object,
    /** Handler for Change event
     * @function
     * @param {string} value
     */
    onChange: PropTypes.func
  }

  static defaultProps = {
    includeMaskChars: false,
    value: ''
  }

  /**
   * Creates Mask meta data for transforming and validating input
   * and placeholder from the pattern if no placeholder is specified
   *
   * @returns {{maskMetadata: Array<Object>, placeholder: string}}
   */
  computePlaceholderAndMaskMetadata = memoize(
    ({ pattern, masks: additionalMasks, placeholder: defaultPlaceholder }) => {
      const masks = { ...defaultMasks, ...additionalMasks }
      const maskMetadata = []
      let placeholder = ''
      if (pattern) {
        for (let maskIndex = 0; maskIndex < pattern.length; ) {
          if (pattern.charAt(maskIndex) === '"') {
            const fixedMask = pattern.slice(maskIndex + 1).split('"')[0]
            placeholder += fixedMask
            maskMetadata.push(
              ...map(fixedMask, maskChar => ({ fixedChar: maskChar, validate: char => char === maskChar }))
            )
            maskIndex += fixedMask.length + 2
          } else {
            maskMetadata.push(masks[pattern.charAt(maskIndex)])
            placeholder += pattern.charAt(maskIndex)
            maskIndex += 1
          }
        }
      }
      return { maskMetadata, placeholder: defaultPlaceholder || placeholder }
    }
  )

  get placeholder() {
    return this.computePlaceholderAndMaskMetadata(this.props).placeholder
  }

  get maskMetadata() {
    return this.computePlaceholderAndMaskMetadata(this.props).maskMetadata
  }

  propagateChange = value => {
    this.props.onChange && this.props.onChange(this.getTargetValue(value))
  }

  /**
   * Creates new placeholder based on current masked input
   * @param {String} maskedValue
   * @returns {String} placeholder
   */
  getPlaceholder = maskedValue => {
    const placeholder = this.placeholder
    const replaceRegex = new RegExp(`^.{${maskedValue.length}}`)
    return placeholder.replace(replaceRegex, maskedValue)
  }

  handleKeyDown = event => {
    switch (event.key || event.keyCode) {
      case 'Backspace':
      case 8:
        event.preventDefault()
        event.stopPropagation()
        const value = this.removeLastChar(event.target.value)
        this.propagateChange(value)
        break
    }
  }

  handleKeyPress = event => {
    const value = event.target.value
    const selectionStart = event.target.selectionStart
    const inputChar = this.transformValue(event.key, selectionStart)

    if (selectionStart === value.length && this.validateValue(inputChar, selectionStart)) {
      const newValue = value + inputChar
      const maskedValue = newValue + this.getNextMask(newValue)
      this.propagateChange(maskedValue)
    }
  }

  transformValue = (char, i) => {
    const mask = this.maskMetadata[i]
    return mask && mask.transform ? mask.transform(char) : char
  }

  validateValue = (char, i) => {
    const mask = this.maskMetadata[i]
    return mask && mask.validate && mask.validate(char)
  }

  /**
   * Get maskedValue from the value prop supplied to the component
   *
   * @param {String} value
   * @returns {String} maskedValue
   */
  getMaskedValue = value => {
    let maskedValue = ''
    const maskMetadata = this.maskMetadata
    if (this.props.includeMaskChars) {
      maskedValue += value
    } else {
      let maskIndex = 0
      for (let valueIndex = 0; valueIndex < value.length; ) {
        if (maskMetadata[maskIndex].fixedChar) {
          maskedValue += maskMetadata[maskIndex].fixedChar
        } else {
          maskedValue += value.charAt(valueIndex)
          valueIndex += 1
        }
        maskIndex += 1
      }
    }
    return this.getValidMaskedValue(maskedValue)
  }

  /**
   * Validates the masked input based on mask definitions
   *
   * @param {String} maskedValue
   * @returns {String} validMaskedValue
   */
  getValidMaskedValue = maskedValue => {
    let validMaskedValue = ''
    for (let valueIndex = 0; valueIndex < maskedValue.length; valueIndex++) {
      const nextChar = this.transformValue(maskedValue.charAt(valueIndex), valueIndex)
      if (!this.validateValue(nextChar, valueIndex)) break
      validMaskedValue += nextChar
    }
    return validMaskedValue + this.getNextMask(validMaskedValue)
  }

  /**
   * Value sent to onchange handler based on whether mask characters are included or not
   *
   * @param {String} value
   * @returns {String} targetValue
   */
  getTargetValue = value => {
    let unmaskedValue = ''
    const maskMetadata = this.maskMetadata
    if (this.props.includeMaskChars) {
      return value
    }
    each(value, (char, i) => {
      unmaskedValue += maskMetadata[i].fixedChar ? '' : char
    })
    return unmaskedValue
  }

  /**
   * Returns the next set of mask characters after current input
   *
   * @param {String} value
   * @return {String} nextMask
   */
  getNextMask = value => {
    const start = value.length
    const maskMetadata = this.maskMetadata
    if (!(maskMetadata[start] && maskMetadata[start].fixedChar)) {
      return ''
    }
    const subMask = maskMetadata.slice(start)
    const end = findIndex(subMask, mask => !mask.fixedChar)
    return subMask
      .slice(0, end < 0 ? subMask.length : end)
      .map(mask => mask.fixedChar)
      .join('')
  }

  /**
   * Removes last character (including mask)
   *
   * @param {String} value
   * @returns {String} valueWithoutLastCharacter
   */
  removeLastChar = value => {
    if (!value.length) return
    const maskMetadata = this.maskMetadata
    if (!maskMetadata[value.length - 1].fixedChar) {
      return value.slice(0, value.length - 1)
    }
    const subMask = maskMetadata.slice(0, value.length)
    const maskStartIndex = findLastIndex(subMask, mask => !mask.fixedChar)
    return value.slice(0, maskStartIndex < 0 ? 0 : maskStartIndex)
  }

  render() {
    const maskedValue = this.getMaskedValue(this.props.value)
    const placeholder = this.getPlaceholder(maskedValue)
    return (
      <div className={classnames('input').use(styles)}>
        <input
          className={classnames('masked-input').use(styles)}
          value={maskedValue}
          onKeyPress={this.handleKeyPress}
          onKeyDown={this.handleKeyDown}
          maxLength={this.maskMetadata.length}
        />
        <input className={classnames('mask').use(styles)} value={placeholder} readOnly />
      </div>
    )
  }
}
