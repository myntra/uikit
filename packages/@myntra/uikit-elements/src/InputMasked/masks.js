export default {
  d: {
    validate: char => /^\d$/.test(char)
  },
  w: {
    validate: char => /^\w$/.test(char)
  },
  L: {
    validate: char => /^[A-Z]$/.test(char),
    transform: char => char.toUpperCase()
  },
  l: {
    validate: char => /^[a-z]$/.test(char),
    transform: char => char.toLowerCase()
  },
  A: {
    validate: char => /^[0-9A-Z_]$/.test(char),
    transform: char => char.toUpperCase()
  },
  a: {
    validate: char => /^[0-9a-z_]$/.test(char),
    transform: char => char.toLowerCase()
  },
  '*': {
    validate: char => /^[a-zA-Z]$/.test(char)
  }
}
