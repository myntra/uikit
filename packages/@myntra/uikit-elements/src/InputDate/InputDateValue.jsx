import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'
import { InputMasked } from '../'

import styles from './InputDateValue.css'
import { parse } from './InputDateUtils'
import Icon from '../Icon/Icon'

const MASKS = {
  Y: {
    validate(char, text) {
      return /\d/.test(char)
    }
  },
  M: {
    validate(char, text) {
      return /\d/.test(char)
    }
  },
  D: {
    validate(char, text) {
      return /\d/.test(char)
    }
  }
}

/**
 {describe component}

 @since 0.0.0
 @status EXPERIMENTAL
 @example
 <InputDate.Value />
 */
class InputDateValue extends PureComponent {
  static propTypes = {
    /** @private */
    className: PropTypes.string,
    range: PropTypes.bool,
    active: PropTypes.oneOf(['from', 'to']),
    ignoreBlur: PropTypes.bool,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.shape({
        from: PropTypes.string,
        to: PropTypes.string
      })
    ]),
    format: PropTypes.string.isRequired,
    /**
     * @function
     * @param {string|{from?: string, to?: string}} value
     */
    onChange: PropTypes.func.isRequired,
    onRangeFocus: PropTypes.func,
    onFocus: PropTypes.func
  }

  state = { value: null }

  get value() {
    return this.state.value ? this.state.value : this.props.value
  }

  get pattern() {
    return this.props.format.toUpperCase().replace(/[^YMD]+/g, match => `"${match}"`)
  }

  handleChange = value => {
    try {
      const date = parse(value, this.props.format)

      this.setState({ value: null })
      this.props.onChange(date)
    } catch (e) {
      this.setState({ value })
    }
  }

  handleFromChange = value => {
    try {
      const date = parse(value, this.props.format)

      this.setState({ value: null })
      this.props.onChange({ from: date })
    } catch (e) {
      this.setState({ value: { from: value, to: this.value.to } })
    }
  }

  handleToChange = value => {
    try {
      const date = parse(value, this.props.format)

      this.setState({ value: null })
      this.props.onChange({ to: date })
    } catch (e) {
      this.setState({ value: { from: this.value.from, to: value } })
    }
  }

  /**
   * @public
   */
  handleBlur = () => {
    this.setState({ value: null })
  }

  handleFromFocus = () => {
    this.handleRangeFocus('from')
  }

  handleToFocus = () => {
    this.handleRangeFocus('to')
  }

  handleRangeFocus = value => {
    this.props.onRangeFocus && this.props.onRangeFocus(value)
    this.props.onFocus && this.props.onFocus()
  }

  handleFromClear = () => {
    this.setState({ value: null })
    this.props.onChange && this.props.onChange({ from: null })
  }

  handleToClear = () => {
    this.setState({ value: null })
    this.props.onChange && this.props.onChange({ to: null })
  }

  handleClear = () => {
    this.setState({ value: null })
    this.props.onChange && this.props.onChange(null)
  }

  render() {
    const props = {
      onBlur: this.handleBlur,
      onFocus: this.props.onFocus,
      masks: MASKS,
      pattern: this.pattern,
      includeMaskChars: true
    }

    return (
      <div className={classnames('date-value').use(styles)}>
        {this.props.range && (
          <div key="from" className={classnames('wrapper').use(styles)}>
            <InputMasked
              {...props}
              className={classnames({ active: this.props.active === 'from' }).use(styles)}
              value={this.value.from}
              onClick={this.handleFromFocus}
              onFocus={this.handleFromFocus}
              onChange={this.handleFromChange}
            />
            {this.props.value.from && (
              <Icon
                className={classnames('icon').use(styles)}
                name="cross"
                title="Clear date"
                onClick={this.handleFromClear}
              />
            )}
          </div>
        )}
        {this.props.range && (
          <div key="to" className={classnames('wrapper').use(styles)}>
            <InputMasked
              {...props}
              className={classnames({ active: this.props.active === 'to' }).use(styles)}
              value={this.value.to}
              onClick={this.handleToFocus}
              onFocus={this.handleToFocus}
              onChange={this.handleToChange}
            />
            {this.props.value.to && (
              <Icon
                className={classnames('icon').use(styles)}
                name="cross"
                title="Clear date"
                onClick={this.handleToClear}
              />
            )}
          </div>
        )}
        {!this.props.range && (
          <div key="both" className={classnames('wrapper').use(styles)}>
            <InputMasked {...props} value={this.value} onChange={this.handleChange} />
            {this.props.value && (
              <Icon
                className={classnames('icon').use(styles)}
                name="cross"
                title="Clear date"
                onClick={this.handleClear}
              />
            )}
          </div>
        )}
      </div>
    )
  }
}

export default InputDateValue
