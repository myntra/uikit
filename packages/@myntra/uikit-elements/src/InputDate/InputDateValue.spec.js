import React from 'react'
import { shallow } from 'enzyme'
import InputDateValue, { MASKS } from './InputDateValue'
import InputMasked from '../InputMasked/InputMasked'
import { Icon } from '..'
import { UTCDate } from './InputDateUtils'

const DATE = UTCDate(2018, 4, 5)
describe('render', () => {
  const date = '2018-05-05'

  test('single', () => {
    const fn = jest.fn()
    const wrapper = shallow(<InputDateValue format="yyyy-MM-dd" value={date} onChange={fn} />)

    expect(wrapper.find(InputMasked)).toHaveLength(1)
    expect(wrapper).toMatchSnapshot()

    wrapper.find(Icon).simulate('click')
    expect(fn).toHaveBeenCalledWith(null)
    fn.mockClear()

    wrapper.find(InputMasked).simulate('change', '2018')

    expect(wrapper.state().value).toBe('2018')
    fn.mockClear()

    wrapper.find(InputMasked).simulate('change', date)

    expect(wrapper.state().value).toBe(null)
    expect(fn).toHaveBeenCalledWith(DATE)
    fn.mockClear()
  })

  test('range', () => {
    const fn = jest.fn()
    const wrapper = shallow(<InputDateValue format="yyyy-MM-dd" onChange={fn} range />)

    expect(wrapper.find(InputMasked)).toHaveLength(2)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('change date', () => {
  const fn = jest.fn()
  const date = '2018-05-05'
  const wrapper = shallow(
    <InputDateValue format="yyyy-MM-dd" onChange={fn} value={{ from: '2018-05-05', to: '2018-05-05' }} range />
  )

  afterEach(() => {
    fn.mockClear()
  })

  it('should render two input masked fields with cross icon', () => {
    expect(wrapper.find(Icon)).toHaveLength(2)
    expect(wrapper.find(InputMasked)).toHaveLength(2)
  })

  it('should clear from date', () => {
    wrapper
      .find(Icon)
      .at(0)
      .simulate('click')
    expect(fn).toHaveBeenCalledWith({ from: null })
  })

  it('should allow typing from date', () => {
    wrapper
      .find(InputMasked)
      .at(0)
      .simulate('change', '2018')
    expect(wrapper.state().value.from).toBe('2018')

    wrapper
      .find(InputMasked)
      .at(0)
      .simulate('blur')
    expect(wrapper.state().value).toBe(null)
  })

  it('should locally store partial changes', () => {
    const wrapper = shallow(<InputDateValue format="MM/dd" onChange={fn} value={{ from: '05/05' }} range />)

    wrapper
      .find(InputMasked)
      .at(0)
      .simulate('change', '12/1')
    expect(wrapper.state().value).toEqual({ from: '12/1' })
  })

  it('should change from date', () => {
    wrapper
      .find(InputMasked)
      .at(0)
      .simulate('change', date)
    expect(wrapper.state().value).toBe(null)
    expect(fn).toHaveBeenCalledWith({ from: DATE })
  })

  it('should clear to date', () => {
    wrapper
      .find(Icon)
      .at(1)
      .simulate('click')
    expect(fn).toHaveBeenCalledWith({ to: null })
  })

  it('should allow typing to date', () => {
    wrapper
      .find(InputMasked)
      .at(1)
      .simulate('change', '2018')
    expect(wrapper.state().value.to).toBe('2018')

    wrapper
      .find(InputMasked)
      .at(1)
      .simulate('blur')
    expect(wrapper.state().value).toBe(null)
  })

  it('should change to date', () => {
    wrapper
      .find(InputMasked)
      .at(1)
      .simulate('change', date)
    expect(wrapper.state().value).toBe(null)
    expect(fn).toHaveBeenCalledWith({ to: DATE })
  })
})

describe('focus from/to date', () => {
  const handleFocus = jest.fn()
  const handleRangeFocus = jest.fn()
  const wrapper = shallow(
    <InputDateValue
      format="yyyy-MM-dd"
      onFocus={handleFocus}
      onRangeFocus={handleRangeFocus}
      value={{ from: '2018-05-05', to: '2018-05-05' }}
      range
      onChange={() => {}}
    />
  )

  afterEach(() => {
    handleFocus.mockClear()
    handleRangeFocus.mockClear()
  })

  test('focus from', () => {
    wrapper
      .find(InputMasked)
      .at(0)
      .simulate('focus')

    expect(handleFocus).toHaveBeenCalled()
    expect(handleRangeFocus).toHaveBeenCalledWith('from')
  })

  test('focus to', () => {
    wrapper
      .find(InputMasked)
      .at(1)
      .simulate('focus')

    expect(handleFocus).toHaveBeenCalled()
    expect(handleRangeFocus).toHaveBeenCalledWith('to')
  })

  test('click from', () => {
    wrapper
      .find(InputMasked)
      .at(0)
      .simulate('click')

    expect(handleFocus).toHaveBeenCalled()
    expect(handleRangeFocus).toHaveBeenCalledWith('from')
  })

  test('click to', () => {
    wrapper
      .find(InputMasked)
      .at(1)
      .simulate('click')

    expect(handleFocus).toHaveBeenCalled()
    expect(handleRangeFocus).toHaveBeenCalledWith('to')
  })
})

describe('MASKS', () => {
  // TODO: Fix masks.
  test('Y', () => {
    expect(MASKS.Y.validate('4')).toBe(true)
    expect(MASKS.Y.validate('z')).toBe(false)
  })
  test('M', () => {
    expect(MASKS.M.validate('4')).toBe(true)
    expect(MASKS.M.validate('z')).toBe(false)
  })
  test('D', () => {
    expect(MASKS.D.validate('4')).toBe(true)
    expect(MASKS.D.validate('z')).toBe(false)
  })
})
