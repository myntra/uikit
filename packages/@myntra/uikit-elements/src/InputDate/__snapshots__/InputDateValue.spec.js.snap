// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render range 1`] = `
<div
  className="date-value"
>
  <div
    className="wrapper"
    key="from"
  >
    <InputMasked
      id="from"
      includeMaskChars={true}
      masks={
        Object {
          "D": Object {
            "validate": [Function],
          },
          "M": Object {
            "validate": [Function],
          },
          "Y": Object {
            "validate": [Function],
          },
        }
      }
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      pattern="YYYY\\"-\\"MM\\"-\\"DD"
      value=""
    />
  </div>
  <div
    className="wrapper"
    key="to"
  >
    <InputMasked
      id="to"
      includeMaskChars={true}
      masks={
        Object {
          "D": Object {
            "validate": [Function],
          },
          "M": Object {
            "validate": [Function],
          },
          "Y": Object {
            "validate": [Function],
          },
        }
      }
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      pattern="YYYY\\"-\\"MM\\"-\\"DD"
      value=""
    />
  </div>
</div>
`;

exports[`render single 1`] = `
<div
  className="date-value"
>
  <div
    className="wrapper"
    key="both"
  >
    <InputMasked
      includeMaskChars={true}
      masks={
        Object {
          "D": Object {
            "validate": [Function],
          },
          "M": Object {
            "validate": [Function],
          },
          "Y": Object {
            "validate": [Function],
          },
        }
      }
      onBlur={[Function]}
      onChange={[Function]}
      pattern="YYYY\\"-\\"MM\\"-\\"DD"
      value="2018-05-05"
    />
    <Icon
      className="icon"
      name="cross"
      onClick={[Function]}
      title="Clear date"
    />
  </div>
</div>
`;
